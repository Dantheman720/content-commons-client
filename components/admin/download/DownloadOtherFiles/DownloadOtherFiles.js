import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { Item, Loader } from 'semantic-ui-react';
import { getS3Url } from 'lib/utils';
import ApolloError from 'components/errors/ApolloError';

import downloadIcon from 'static/icons/icon_download.svg';

const DownloadOtherFiles = props => {
  const { data, instructions, isPreview } = props;
  const { error, loading, project } = data;

  if ( loading ) {
    return (
      <div style={ {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '200px'
      } }
      >
        <Loader
          active
          inline="centered"
          style={ { marginBottom: '1em' } }
          content="Loading other file(s)..."
        />
      </div>
    );
  }

  if ( error ) return <ApolloError error={ error } />;
  if ( !project || !Object.keys( project ).length ) return null;

  const { files } = project;

  const renderFormItem = file => {
    const {
      id, filename, filetype, url,
      language: { displayName }
    } = file;

    return (
      <Item.Group key={ `fs_${id}` } className={ `download-item${isPreview ? ' preview' : ''}` }>
        <Item
          as={ isPreview ? 'span' : 'a' }
          href={ isPreview ? null : getS3Url( url ) }
          download={ isPreview ? null : filename }
        >
          <Item.Image size="mini" src={ downloadIcon } className="download-icon" />
          <Item.Content>
            <Item.Header className="download-header">
              { `Download ${displayName} ${filetype} file` }
            </Item.Header>
            <span className="item_hover">
              { `Download ${displayName} ${filetype} file` }
              { isPreview
                && (
                  <span className="preview-text">
                    The link will be active after publishing.
                  </span>
                ) }
            </span>
          </Item.Content>
        </Item>
      </Item.Group>
    );
  };

  const renderFormItems = () => {
    const otherFiles = files
      .filter( file => file && file.url )
      .map( file => renderFormItem( file ) );
    return otherFiles.length ? otherFiles : 'There are no other files available for download at this time';
  };

  return (
    <Fragment>
      <p className="form-group_instructions">{ instructions }</p>
      { files && renderFormItems( files ) }
    </Fragment>
  );
};

DownloadOtherFiles.propTypes = {
  data: PropTypes.object,
  instructions: PropTypes.string,
  isPreview: PropTypes.bool
};

const VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY = gql`
  query VideoProjectPreviewOtherfiles($id: ID!) {
    project: videoProject(id: $id) {
      id
      files: supportFiles(
        where: {
          filetype_not: "application/x-subrip"
        },
        orderBy: filename_ASC
      ) {
        id
        filename
        filetype
        url
        language {
          id
          displayName
        }
      }
    }
  }
`;

export default graphql( VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY, {
  options: props => ( {
    variables: {
      id: props.id
    },
  } )
} )( DownloadOtherFiles );

export { VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY };
