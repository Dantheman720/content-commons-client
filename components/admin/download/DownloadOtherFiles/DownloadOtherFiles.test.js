import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Loader } from 'semantic-ui-react';
import DownloadOtherFiles, { VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY } from './DownloadOtherFiles';

jest.mock( 'lib/utils', () => ( {
  getPathToS3Bucket: jest.fn( () => {} ),
  getS3Url: jest.fn( assetPath => (
    `https://s3-url.com/${assetPath}`
  ) )
} ) );

jest.mock( 'static/icons/icon_download.svg', () => 'downloadIconSVG' );

const props = {
  id: '123',
  instructions: 'Download Other File(s)',
  isPreview: false
};

const mocks = [
  {
    request: {
      query: VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY,
      variables: { id: props.id }
    },
    result: {
      data: {
        project: {
          id: props.id,
          files: [
            {
              id: 'pdf89',
              filename: 'file-1.pdf',
              filetype: 'application/pdf',
              url: `2019/06/${props.id}/file-1.pdf`,
              language: {
                id: 'en23',
                displayName: 'English'
              }
            },
            {
              id: 'au56',
              filename: 'audio-1.mp3',
              filetype: 'audio/x-mpeg-3',
              url: `2019/06/${props.id}/audio-1.mp3`,
              language: {
                id: 'en23',
                displayName: 'English'
              }
            }
          ]
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DownloadOtherFiles { ...props } />
  </MockedProvider>
);

describe( '<DownloadOtherFiles />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading other file(s)..."
      />
    );

    expect( downloadOther.exists() ).toEqual( true );
    expect( downloadOther.contains( loader ) ).toEqual( true );
    expect( toJSON( downloadOther ) ).toMatchSnapshot();
  } );

  it( 'renders error message if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY,
          variables: { id: props.id }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const errorComponent = downloadOther.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders final state without crashing', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( toJSON( downloadOther ) ).toMatchSnapshot();
  } );

  it( 'renders null if project is null', async () => {
    const nullProjectMocks = [
      {
        ...mocks[0],
        result: { data: { project: null } }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullProjectMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( downloadOther.html() ).toEqual( null );
  } );

  it( 'renders null if project is {}', async () => {
    // ignore console.warn about missing field `id` and `files`
    const emptyProjectMocks = [
      {
        ...mocks[0],
        result: { data: { project: {} } }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ emptyProjectMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( downloadOther.html() ).toEqual( null );
  } );

  it( 'renders a "no other files available message" if there are no other files', async () => {
    const noFilesMocks = [
      {
        ...mocks[0],
        result: {
          data: {
            project: {
              ...mocks[0].result.data.project,
              files: []
            }
          }
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const items = downloadOther.find( 'Item' );
    const { files } = noFilesMocks[0].result.data.project;
    const msg = 'There are no other files available for download at this time';

    expect( downloadOther.exists() ).toEqual( true );
    expect( items.length ).toEqual( files.length );
    expect( downloadOther.contains( msg ) ).toEqual( true );
  } );

  it( 'renders <a> tags with the correct href and download attribute values', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( 'Item' );
    const { files } = mocks[0].result.data.project;
    const s3Bucket = 'https://s3-url.com';

    expect( items.length ).toEqual( files.length );
    items.forEach( ( item, i ) => {
      const { url: assetPath, filename } = files[i];
      expect( item.prop( 'href' ) ).toEqual( `${s3Bucket}/${assetPath}` );
      expect( item.prop( 'download' ) ).toEqual( filename );
    } );
  } );

  it( 'renders <span> tags with null href & download attributes if isPreview is true', async () => {
    const newProps = { ...props, isPreview: true };
    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <DownloadOtherFiles { ...newProps } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( '.item' );
    const { files } = mocks[0].result.data.project;

    expect( items.length ).toEqual( files.length );
    items.forEach( item => {
      expect( item.name() ).toEqual( 'span' );
      expect( item.prop( 'href' ) ).toEqual( null );
      expect( item.prop( 'download' ) ).toEqual( null );
    } );
  } );
} );
