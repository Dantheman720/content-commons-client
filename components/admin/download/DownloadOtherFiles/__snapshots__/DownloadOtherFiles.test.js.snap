// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadOtherFiles /> renders final state without crashing 1`] = `
<DownloadOtherFiles
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "project": Object {
        "files": Array [
          Object {
            "filename": "file-1.pdf",
            "filetype": "application/pdf",
            "id": "pdf89",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/file-1.pdf",
          },
        ],
        "id": "123",
      },
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download SRT(s)"
>
  <p
    className="form-group_instructions"
  >
    Download SRT(s)
  </p>
  <ItemGroup
    className="download-item"
    key="fs_pdf89"
  >
    <div
      className="ui items download-item"
    >
      <Item
        as="a"
        download="file-1.pdf"
        href="https://s3-url.com/2019/06/123/file-1.pdf"
      >
        <a
          className="item"
          download="file-1.pdf"
          href="https://s3-url.com/2019/06/123/file-1.pdf"
        >
          <ItemImage
            className="download-icon"
            size="mini"
            src="downloadIconSVG"
          >
            <Image
              as="img"
              className="download-icon"
              size="mini"
              src="downloadIconSVG"
              ui={true}
              wrapped={true}
            >
              <div
                className="ui mini image download-icon"
              >
                <img
                  src="downloadIconSVG"
                />
              </div>
            </Image>
          </ItemImage>
          <ItemContent>
            <div
              className="content"
            >
              <ItemHeader
                className="download-header"
              >
                <div
                  className="header download-header"
                >
                  Download English application/pdf file
                </div>
              </ItemHeader>
              <span
                className="item_hover"
              >
                Download English application/pdf file
              </span>
            </div>
          </ItemContent>
        </a>
      </Item>
    </div>
  </ItemGroup>
</DownloadOtherFiles>
`;

exports[`<DownloadOtherFiles /> renders initial loading state without crashing 1`] = `
<DownloadOtherFiles
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download SRT(s)"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      content="Loading other file(s)..."
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active text centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      >
        Loading other file(s)...
      </div>
    </Loader>
  </div>
</DownloadOtherFiles>
`;
